src/instructlab/sdg/generate_data.py

```diff
@@ -318,47 +265,78 @@ def generate_data(
             "Synthesizing new instructions. If you aren't satisfied with the generated instructions, interrupt training (Ctrl-C) and try adjusting your YAML files. Adding more examples may help."
         )
 
-    generated_data = None
-    for leaf_node in leaf_nodes.values():
+    for i, leaf_node in enumerate(leaf_nodes.values()):
+        is_knowledge = False
         samples = leaf_node_to_samples(leaf_node, server_ctx_size, chunk_word_count)
+        ds = Dataset.from_list(samples)
 
         if not samples:
             raise GenerateException("Error: No samples found in leaf node.")
 
         if samples[0].get("document"):
             sdg = sdg_knowledge
+            logger.info(f"Generating data for leaf node {i} with knowledge pipeline.")
+            is_knowledge = True
+            # add to 0.7 recipe
+            # add to 1.0 recipe
+
         elif samples[0].get("seed_context"):
             sdg = sdg_grounded_skill
+            logger.info(
+                f"Generating data for leaf node {i} with grounded skill pipeline."
+            )
+            # add to 1.0 recipe
+
         else:
             sdg = sdg_freeform_skill
+            logger.info(
+                f"Generating data for leaf node {i} with freeform skill pipeline."
+            )
+            # add to 1.0 recipe
 
-        logger.debug("Samples: %s" % samples)
-        ds = Dataset.from_list(samples)
-        logger.debug("Dataset: %s" % ds)
-        new_generated_data = sdg.generate(ds)
-        generated_data = (
-            [new_generated_data]
-            if generated_data is None
-            else generated_data + [new_generated_data]
-        )
-        logger.info("Generated %d samples" % len(generated_data))
-        logger.debug("Generated data: %s" % generated_data)
+        generated_data = sdg.generate(ds)
 
-    if generated_data is None:
-        generated_data = []
+        if is_knowledge:
+            knowledge_phase_data = create_phase07_ds(generated_data)
+            skills_phase_data = create_phase10_ds(generated_data)
 
-    _gen_train_data(
-        logger,
-        generated_data,
-        os.path.join(output_dir, output_file_train),
-        os.path.join(output_dir, output_file_messages),
-    )
+            knowledge_fpath = os.path.join(
+                output_dir, f"node_datasets_{date_suffix}/node_{i}_p07.jsonl"
+            )
+            skills_fpath = os.path.join(
+                output_dir, f"node_datasets_{date_suffix}/node_{i}_p10.jsonl"
+            )
+            knowledge_phase_data.to_json(knowledge_fpath, orient="records", lines=True)
+            skills_phase_data.to_json(skills_fpath, orient="records", lines=True)
+
+            knowledge_recipe.add_dataset(knowledge_fpath)
+            skills_recipe.add_dataset(skills_fpath)
```

Comment:
Is the roundtrip to disk necessary? i.e. we're writing to the dataset to a file just for it to be loaded again shortly after?

The benefit is that we have a record of this intermediate stage? Or reduced memory usage?

Reply:
My understanding is that it's so we have an artifact trail of the raw generated data as well as the mixed data, and when #185 goes in the recipe files will reference this raw generated data to show what was mixed into the overall mixed dataset. The mixing process can reduce the number of samples from a given taxonomy leaf in the overall mixed dataset, so these intermediate files also allow the entire generated output for each leaf node to be retained as an artifact.

Reduce memory usage is the other side-effect of this, since we're only ever keeping the dataset for a single leaf node in memory until the mixing process, where we sample some number/percent of the overall generated samples for each leaf instead of loading the entirety of all leaf node generated samples at once.